### Reduction operation: the sum of the numbers in the range [0, a)

In [8]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n") 

Time taken by reduction operation: 0.27745771408081055 seconds
255 ms ± 5.95 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [9]:
# Ejercicio 3.2:

# a) Uso de listas
lista = list(range(value))

# a.1) Suma usando bucle for
initialTime = time.time()
suma_lista = 0
for i in lista:
    suma_lista += i
finalTime = time.time()
print("Time with list + for:", finalTime - initialTime, "seconds")

# a.2) Suma usando sum()
initialTime = time.time()
suma_builtin = sum(lista)
finalTime = time.time()
print("Time with list + sum():", finalTime - initialTime, "seconds")

# Validación
print("Resultado con sum(lista):", suma_builtin)

# Con timeit
%timeit -r 2 sum(lista)

Time with list + for: 0.4436678886413574 seconds
Time with list + sum(): 0.030950069427490234 seconds
Resultado con sum(lista): 499999500000
29.8 ms ± 1.15 µs per loop (mean ± std. dev. of 2 runs, 10 loops each)


In [11]:
import numpy as np

# b) Conversión a array de numpy
array = np.array(lista)

# b.1) Suma con bucle for
initialTime = time.time()
suma_np_for = 0
for i in array:
    suma_np_for += i
finalTime = time.time()
print("Time with numpy array + for:", finalTime - initialTime, "seconds")

# b.2) np.sum()
initialTime = time.time()
suma_np = np.sum(array)
finalTime = time.time()
print("Time with numpy + np.sum():", finalTime - initialTime, "seconds")
print("Resultado con np.sum:", suma_np)

# Con timeit
%timeit -r 2 np.sum(array)

Time with numpy array + for: 0.9317800998687744 seconds
Time with numpy + np.sum(): 0.0018625259399414062 seconds
Resultado con np.sum: 499999500000
1.37 ms ± 4.89 µs per loop (mean ± std. dev. of 2 runs, 1,000 loops each)


### Comparativa entre Python puro, listas, y NumPy
Los resultados obtenidos han sido los siguientes:
#### a) Listas de Python
- **Bucle for sobre lista:** ~0.44 segundos
- **sum(lista):** ~0.031 segundos
- **%timeit sum(lista):** ~29.8 ms por ejecución
Usar `sum()` directamente sobre la lista es significativamente más rápido que recorrerla manualmente con un bucle `for`. Esto se debe a que `sum()` está implementada en C internamente, lo que la hace mucho más eficiente que un bucle de Python.
#### b) Arrays de NumPy
- **Bucle for sobre array:** ~0.94 segundos
- **np.sum(array):** ~0.0025 segundos
- **%timeit np.sum(array):** ~1.4 ms por ejecución

Aunque los arrays de NumPy ofrecen ventajas para operaciones, recorrerlos manualmente con un bucle `for` es incluso más lento que hacerlo sobre listas.
Sin embargo, el uso de `np.sum()` proporciona el mejor rendimiento de todos los métodos probados. Es ideal para trabajar con grandes volúmenes de datos.